File: data/sorting/bucket-sort.cpp
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

vector<float> bucketSort(vector<float>& nums) {
    int n = nums.size();
    if (n <= 0) return nums;

    vector<vector<float>> buckets(n);

    for (int i = 0; i < n; ++i) {
        int bucketIdx = n * nums[i];
        if (bucketIdx >= n) bucketIdx = n - 1;
        buckets[bucketIdx].push_back(nums[i]);
    }

    for (int i = 0; i < n; ++i) {
        sort(buckets[i].begin(), buckets[i].end());
    }

    int idx = 0;
    for (int i = 0; i < n; ++i) {
        for (float val : buckets[i]) {
            nums[idx++] = val;
        }
    }
    return nums;
}

File: data/sorting/bucket-sort.py
def bucketSort(nums: list[float]) -> list[float]:
    """
    Bucket Sort Algorithm
    Time Complexity: O(n + k)
    Space Complexity: O(n + k)
    """
    if not nums:
        return nums

    bucket_count = len(nums)
    buckets = [[] for _ in range(bucket_count)]

    for num in nums:
        index = int(num * bucket_count)
        if index >= bucket_count:
            index = bucket_count - 1
        buckets[index].append(num)

    for bucket in buckets:
        bucket.sort()

    k = 0
    for bucket in buckets:
        for val in bucket:
            nums[k] = val
            k += 1

    return nums
